### PR TITLE
HPCC-21126 Refactor CXRefExBuilderThread class

### PR DESCRIPTION
**Pull request by wangkx**
1. Fix clang warning in dfuXRefService.hpp;
2. Add another CriticalSection for accessing job queue;
3. Clean the code in the CXRefExBuilderThread class.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by wangkx** [comment]
@afishbeck please review.

File: esp/services/ws_dfu/ws_dfuXRefService.hpp [hunk 1]
```diff
@@ -26,116 +26,135 @@
...
     }
-    virtual bool IsQueued(const char* Queue)
+    virtual bool isQueued(const char* clusterName)
     {
-        if(Queue == 0)
+        if (isEmptyString(clusterName))
             return false;
-        ForEachItemIn(x,m_pNodeQueue)
+
+        if (!currentClusterName.isEmpty() && streq(currentClusterName, clusterName))
```
Comment (afishbeck) [inline esp/services/ws_dfu/ws_dfuXRefService.hpp]: @wangkx is this called from a different thread then runXRef? Should currentClusterName be protected?

**Review by afishbeck — COMMENTED** [review]
@wangkx one comment.

File: esp/services/ws_dfu/ws_dfuXRefService.hpp (hunk 1 above)
Comment (wangkx) [reply esp/services/ws_dfu/ws_dfuXRefService.hpp]: Good point. But, protecting currentClusterName may not fix the problem. Let me think about it.

**Comment by wangkx** [comment]
@afishbeck please review my changes.

**Review by afishbeck — APPROVED** [review]
@wangkx Looks ok now.

File: esp/services/ws_dfu/ws_dfuXRefService.hpp [hunk 2]
```diff
@@ -26,116 +26,144 @@
...
-        m_sem.signal();
-
     }
 
-    virtual bool IsRunning()
+    virtual bool isRunning()
     {
-        CriticalSection(_boolMutex);
-        return bRunning;
+        CriticalBlock b(critRunningStatus);
```
Comment (richardkchapman) [inline esp/services/ws_dfu/ws_dfuXRefService.hpp]: Bit pointless using a CriticalBlock here, I think. If there's a race condition where someone can be changing this value around the same time as you are calling it, you are still going to get a random value.

**Comment by richardkchapman** [comment]
@ghalliday Want to take a quick look?

File: esp/services/ws_dfu/ws_dfuXRefService.hpp [hunk 3]
```diff
@@ -26,116 +26,144 @@
...
-       CriticalSection(_RunningMutex);
-       
-       xRefNode->setCluster(cluster);
-       xRefNode->Link();
-       m_pNodeQueue.enqueue(xRefNode);
-       m_sem.signal();
-   }
-
-   ~CXRefExBuilderThread(){DBGLOG("Destroyed XRef thread");}
+    IMPLEMENT_IINTERFACE;
```
Comment (ghalliday) [inline esp/services/ws_dfu/ws_dfuXRefService.hpp]: minor: Not needed.

**Review by ghalliday — COMMENTED** [review]
@richardkchapman one small comment.  I agree with your comment above, but the code should work as it stands.

**Pull request merged**